File: clients/android/app/src/main/java/eu/allodia/mailcal/CalendarTrace.kt
// What the grid did, in numbers, the instrument for the bugs that only exist in a hand.
//
// It went in as temporary scaffolding and stayed, because it found two things nothing else could.
// A pinch frame costing 3.4x a swipe frame *while drawing half the blocks* is not "the pinch is
// heavy", cost going up as work goes down is a clue, and it pointed straight at the text shaper
// missing its cache on every frame of a zoom. And logging what the single owner DECIDED each finger
// was (`pan_x` / `pan_y` / `zoom`) is the only way to prove a real hand's pinch is not being misread
// as a pan: `adb` cannot inject two fingers, so no script will ever tell you.
//
// **This logs counts and durations. It never logs an event.** No title, no time, no attendee, no
// calendar name, the developer validating this has their own mail and their own diary on the device,
// and docs/logging.md's never-log-content rule is absolute regardless of how convenient a title would
// be in a trace. What a frame *costs* is a number; what is *in* it is nobody's business.
//
// Off unless asked for, in **any** build including release, which is the point, because the only
// build worth judging the grid on is a release one (an unminified Compose build is several times
// slower, and that is exactly how the old grid came to be measured against Samsung's release build
// while running as a debug one). So the switch is a log tag rather than a debug flag, and the cost of
// being switched off is one cached boolean:
//
//     adb shell setprop log.tag.MailcalCal DEBUG     # then launch
//     adb logcat -s MailcalCal
//
// Read once, into a `val`: the per-frame cost of being switched off is one boolean.
package eu.allodia.mailcal

import android.util.Log

private const val TAG = "MailcalCal"

/**
 * What the grid did, in numbers, flushed once a second.
 *
 * Deliberately not a `Logger` port call per event: the point is to measure a frame, and a trace that
 * writes to disk inside the frame it is measuring is measuring itself.
 */
internal object CalendarTrace {
    /** Cached at class-load. `setprop` before launching, not after. */
    @JvmField val on: Boolean = Log.isLoggable(TAG, Log.DEBUG)

    private var frames = 0
    private var blocksDrawn = 0
    private var blocksCulled = 0
    private var measures = 0
    private var drawNanos = 0L
    private var worstDrawNanos = 0L
    private var pagesPainted = 0
    private var panDays = 0
    private var panHours = 0
    private var zooms = 0
    private var taps = 0
    private var turns = 0
    private var lastFlush = 0L

    /** One page, drawn. [culled] is what the viewport saved us, the whole point of drawing by hand. */
    fun frame(drawn: Int, culled: Int, nanos: Long) {
        if (!on) return
        frames++
        blocksDrawn += drawn
        blocksCulled += culled
        drawNanos += nanos
        if (nanos > worstDrawNanos) worstDrawNanos = nanos
        flush(System.nanoTime())
    }

    /** A line of text shaped. A pinch that misses the shaper's cache does this on every frame. */
    fun measured() {
        if (on) measures++
    }

    /** A week turned into a [PagePaint], should be *one* per swipe, not three. */
    fun painted() {
        if (on) pagesPainted++
    }

    /**
     * What the single gesture owner decided this finger was.
     *
     * Flushes on its own clock as well as the frame's. It used to flush only from [frame], and so a
     * burst of flicks that ended with the grid going idle left its last counters in limbo, the log
     * reported six of eight gestures, and the two it swallowed were its own, not the app's. An
     * instrument that under-reports is worse than no instrument, because it invents a bug.
     */
    fun gesture(mode: String) {
        if (!on) return
        when (mode) {
            "days" -> panDays++
            "hours" -> panHours++
            "zoom" -> zooms++
            "tap" -> taps++
        }
        flush(System.nanoTime())
    }

    /** A week committed. */
    fun turned() {
        if (on) turns++
    }

    /** A settled pinch, with the rung it landed on, the numbers, not the gesture. */
    fun settled(hours: Int, columns: Int) {
        if (on) Log.d(TAG, "zoom settled: horizon=${hours}h columns=$columns")
    }

    private fun flush(now: Long) {
        if (lastFlush == 0L) lastFlush = now
        val elapsed = now - lastFlush
        if (elapsed < 1_000_000_000L) return

        val avgDrawUs = if (frames > 0) drawNanos / frames / 1_000 else 0
        val worstUs = worstDrawNanos / 1_000
        Log.d(
            TAG,
            "1s: frames=$frames draw_avg=${avgDrawUs}us draw_worst=${worstUs}us " +
                "blocks=$blocksDrawn culled=$blocksCulled measures=$measures " +
                "paints=$pagesPainted | pan_x=$panDays pan_y=$panHours zoom=$zooms tap=$taps " +
                "turns=$turns",
        )
        frames = 0
        blocksDrawn = 0
        blocksCulled = 0
        measures = 0
        drawNanos = 0
        worstDrawNanos = 0
        pagesPainted = 0
        panDays = 0
        panHours = 0
        zooms = 0
        taps = 0
        turns = 0
        lastFlush = now
    }
}

File: clients/windows/Mailcal/Calendar/CalendarTrace.cs
// What the grid decided each finger was, and what each frame cost, counts and durations, never
// content.
//
// Off in every build unless asked for, and it costs one cached boolean. It reports what the single
// gesture owner DECIDED, pan / page / zoom / tap, which is how you catch a pinch being misread as a
// pan by a real hand, and it is what proves the zoom no longer pans the grid.
//
// It found Android's shaper bug, too: a pinch frame costing 3.4x a swipe frame while drawing HALF the
// blocks is not a performance problem, it is a clue.
//
// **It never logs a title, a time, or an attendee.** This runs against real diaries, and
// docs/logging.md's never-log-content rule does not bend for convenience.
using System;
using System.Diagnostics;
using Allodia.Mailcal.Services;

namespace Allodia.Mailcal.Calendar;

/// <summary>The grid's diagnostic counters. Enabled by <c>MAILCAL_CAL_TRACE=1</c>.</summary>
internal sealed class CalendarTrace
{
    /// <summary>How often to report, in frames. Roughly once a second while something is moving.</summary>
    private const int ReportEvery = 60;

    private static readonly bool Enabled =
        Environment.GetEnvironmentVariable("MAILCAL_CAL_TRACE") is "1" or "true";

    private readonly Stopwatch _frame = new();
    private int _frames;
    private int _drawn;
    private int _culled;
    private int _painted;
    private int _turned;
    private double _frameMsTotal;
    private double _frameMsWorst;

    /// <summary>Whether the trace is on. Check it before doing any work for it.</summary>
    internal static bool On => Enabled;

    /// <summary>A page was rebuilt from the core, the thing a zoom must never cause.</summary>
    internal void Painted()
    {
        if (Enabled)
        {
            _painted++;
        }
    }

    /// <summary>A week was banked.</summary>
    internal void Turned()
    {
        if (Enabled)
        {
            _turned++;
        }
    }

    /// <summary>A block was drawn.</summary>
    internal void Drew()
    {
        if (Enabled)
        {
            _drawn++;
        }
    }

    /// <summary>A block was culled, outside the viewport, so it cost one comparison.</summary>
    internal void Culled()
    {
        if (Enabled)
        {
            _culled++;
        }
    }

    /// <summary>What the single owner decided this gesture was.</summary>
    internal static void Gesture(GestureMode mode)
    {
        if (Enabled)
        {
            Log.Info($"cal gesture={mode}");
        }
    }

    internal void FrameBegin()
    {
        if (Enabled)
        {
            _frame.Restart();
        }
    }

    /// <summary>
    /// Ends a frame and, once a second, reports what the last sixty cost.
    /// </summary>
    /// <remarks>
    /// The <b>worst</b> frame is reported alongside the mean, deliberately. A mean hides exactly the
    /// thing the eye sees: the grid is judged on the frames it MISSES during motion, not on the ones
    /// it makes (§7).
    /// </remarks>
    internal void FrameEnd(int shaped)
    {
        if (!Enabled)
        {
            return;
        }
        _frame.Stop();
        var ms = _frame.Elapsed.TotalMilliseconds;
        _frames++;
        _frameMsTotal += ms;
        _frameMsWorst = Math.Max(_frameMsWorst, ms);

        if (_frames < ReportEvery)
        {
            return;
        }
        Log.Info(
            $"cal frames={_frames} mean={_frameMsTotal / _frames:F2}ms worst={_frameMsWorst:F2}ms " +
            $"blocks={_drawn} culled={_culled} shaped={shaped} pages={_painted} turns={_turned}");
        _frames = 0;
        _drawn = 0;
        _culled = 0;
        _painted = 0;
        _turned = 0;
        _frameMsTotal = 0;
        _frameMsWorst = 0;
    }
}
